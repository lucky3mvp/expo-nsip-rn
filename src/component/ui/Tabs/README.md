---
title: Tabs
subTitle: 标签页
---

# Tab 标签页
---
### 组件描述
用于让用户在不同的视图中进行切换。

### 示例代码
```js
const TabPane = Tabs.TabPane;

// 使用 defaultActiveKey
<Tabs defaultActiveKey="tab1">
	<TabPane tab="tab1" key="tab1">
		<View>
			<Text>tab1</Text>
		</View>
	</TabPane>
	<TabPane tab="tab2" key="tab2">
		<View>
			<Text>tab2</Text>
		</View>
	</TabPane>
	<TabPane tab="tab3" key="tab3">
		<View>
			<Text>tab3</Text>
		</View>
	</TabPane>
</Tabs>

// 使用 activeKey, 需配合 onChange 使用
<Tabs activeKey={this.state.activeKey} onChange={key=>this.setSatte({activeKey: key})}>
	<TabPane tab="tab1" key="tab1">
		<View>
			<Text>tab1</Text>
		</View>
	</TabPane>
	<TabPane tab="tab2" key="tab2">
		<View>
			<Text>tab2</Text>
		</View>
	</TabPane>
	<TabPane tab="tab3" key="tab3">
		<View>
			<Text>tab3</Text>
		</View>
	</TabPane>
</Tabs>
```

## API

###  props列表

属性 | 说明 | 类型 | 默认值
----|-----|------|------
| activeKey | 受控属性，当前激活 tab 面板的 key，需配合 onChange 维护 activeKey 值 | string/number | 无 |
| defaultActiveKey | 非受控属性，默认激活的 tab 面板的 key | string/number | 第一个面板 |
| renderTabBar | 是否渲染 tabBar | bool | true |
| tabBarPosition | tab 位置, 取值范围为：top, bottom | string | top |
| onChange | 切换面板的回调，参数为点击的面板的 key, tab 值 | function(key, tab){} | 无 |
| onTabClick | 点击 tab 的回调，参数为点击的面板的 key, tab 值 | function(key, tab){} | 无 |
| animated | 是否显示切换动画 | bool | true |
| swipeable | 是否可以滑动 tab 内容进行切换 | bool| true |
| compose | 是否对内容进行压缩, 为 true 的话，当前未激活过的面板内容将以 <View /> 的形式存在。请谨慎使用，因为为 true 时，可能导致 tab 切换时存在视觉上的断层现象 | bool | false |
| preRenderNum | 预加载的数量，默认值 0，及当 props 或 state 变化时，会更新重绘当前显示的面板内容，设为值 n 后，当 props 或 state 变化时，会重新渲染当前面板以及当前面板左右 n 范围内的面板 | number | 0 |

### 样式对象styles

属性 | 说明 | 适用类型
----|-----|------
| tab | 单个 tab 标签样式 | View |
| text | tab 标签内文字样式 | Text |
| activeText | 激活态 tab 标签内文字样式 | Text |
| bar | 标签栏样式，tab 的容器 | View |
| barTop | 标签栏在 top 时的样式 | View |
| barBottom | 标签栏在 bottom 时的样式 | View |
| activeUnderline | 激活态的 tab 下划线的样式，注意，下划线的颜色实际是通过 View 的 backgroundColor 属性来进行填充的 | View |

### 简单样式 simpleStyle

属性 | 说明 | 适用类型
----|-----|------
| textColor | tab 标签内文字颜色 | Text |
| activeTextColor | 激活态 tab 标签内文字颜色 | Text |
| underlineColor | 标签栏整体下划线的颜色 | View |
| activeUnderlineColor | 激活态的 tab 下划线的颜色，注意，下划线的颜色实际是通过 View 的 backgroundColor 属性来进行填充的 | View |


### Tabs.TabPane

属性 | 说明 | 类型 | 默认值
----|-----|------|------
| key  | tabPane key | string/number  | 无 |
| tab  | 选项卡头显示文字 | string | 无 |

### tabbar 宽度说明
+ tabBar 默认撑满屏幕宽，各个 tab 的宽度则是对屏幕宽的均分
+ 可以通过指定 tabBar 的宽度来改变默认占满屏幕宽的规则，此时
	+ 若 tabBar 宽 < 屏幕宽，则其宽度不会撑满屏幕宽
	+ 若 tabBar 宽 > 屏幕宽，则其宽度会超过屏幕宽而导致生成滚动的 tabBar 