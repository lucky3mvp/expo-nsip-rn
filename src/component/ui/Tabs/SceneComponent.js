import React from 'react';
import { View, StyleSheet } from 'react-native';
import StaticContainer from 'react-static-container';

export default SceneComponent = (Props) => {
	const {shouldUpdated, ...props, } = Props;
	return (
		<View {...props}>
			<StaticContainer shouldUpdate={shouldUpdated}>
				{props.children}
			</StaticContainer>
		</View>
	);
};