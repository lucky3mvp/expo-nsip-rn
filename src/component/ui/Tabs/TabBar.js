import React, { Component } from 'react';
import PropTypes from 'prop-types'
import {
	Text,
	View,
	Animated,
	ScrollView,
	StyleSheet,
	TouchableWithoutFeedback
} from 'react-native';

export default class DefaultTabBar extends Component {
	static defaultProps = {
		onTabClick: () => {},
		onScrollEnd: () => {},
		onAnimationEnd: () => {},
	}

	static propTypes = {
		// tab 标签
		tabs: PropTypes.array,
		// 当前激活的 tab 在 tabs 数组中的下标
		activeIdx: PropTypes.number,
		// animated value, 用户控制下划线的位置
		aniScrollIdx: PropTypes.object, // animated value
		// tabBar 容器宽
		containerWidth: PropTypes.number,
		// 内容区宽
		viewWidth: PropTypes.number,
		// 点击 tab 标签的回调
		onTabClick: PropTypes.func,
		// tabBar 可滚动时，滚动后的回调，参数为滚动后的 x 方向上的位移
		onScrollEnd: PropTypes.func,
		// tabBar 位置更新后的回调
		onAnimationEnd: PropTypes.func,
		// tabBar 容器样式
		style: View.propTypes.style,
		// tab 标签样式
		tabStyle: View.propTypes.style,
		// tab 文字样式
		textStyle: Text.propTypes.style,
		// 激活态文字样式
		activeTextStyle: Text.propTypes.style,
		// 激活态下划线样式
		activeUnderlineStyle: View.propTypes.style,
	}

	constructor(props) {
		super(props)

		this._itemWidth = this._getItemWidth(props);
		this.state = {
			aniLeft: new Animated.Value(props.activeIdx * this._itemWidth)
		}
	}

	componentWillReceiveProps(nextProps) {
		let index = nextProps.activeIdx;

		if (index !== this.props.activeIdx) {
			this._itemWidth = this._getItemWidth(nextProps);

			if (this._handler) {
				this._handler.stop()
				this._handler = null
			}
			
			this._handler = Animated.timing(
				this.state.aniLeft,
				{
					toValue: index * this._itemWidth,
					duration: 350
				}
			).start( () => {
				console.log('tabbar scroll end')
				this._handler = null;
				this.props.onAnimationEnd(index)
			})
		}
	}

	_getItemWidth(props) {
		let { containerWidth, viewWidth, tabs } = props;

		return (containerWidth ? containerWidth : viewWidth) / tabs.length;
	}

	_onTabClick = (tab) => {
		this.props.onTabClick(tab._key, tab.label, tab.index) 
	}

	// _onScroll = (e) => {
	// }
	// _onMomentumScrollBegin = (e) => {
	// }
	// _onMomentumScrollEnd = (e) => {
	// }
	// _onScrollBeginDrag = (e) => {
	// }

	_onScrollEndDrag = (e) => {
		this.props.onScrollEnd(e.nativeEvent.contentOffset.x);
	}

	renderTab(tab, activeIdx) {
		let isTabActive = activeIdx === tab.index,
			{ textStyle, activeTextStyle } = this.props,
			_activeTextStyle = isTabActive ? activeTextStyle : {},
			fontWeight = isTabActive ? 'bold' : 'normal';

		return (
			<TouchableWithoutFeedback
				style={{flex: 1}}
				key={tab.index}
				accessible={true}
				accessibilityLabel={tab.label}
				accessibilityTraits='button'
				onPress={this._onTabClick.bind(this, tab)}
			>
				<View style={this.props.tabStyle}>
					<Text style={[textStyle, _activeTextStyle, { fontWeight }]}>{tab.label}</Text>
				</View>
			</TouchableWithoutFeedback>
		);
	}

	render() {
		let { containerWidth, viewWidth, tabs, style, activeIdx, activeUnderlineStyle, aniScrollIdx } = this.props,
			width = containerWidth ? containerWidth : viewWidth;
		
		// let left = aniScrollIdx.interpolate({
		// 	inputRange: [0, 1], outputRange: [0, this._itemWidth],
		// });

		let cnt = (
			<View style={style}>
				{tabs.map( tab => this.renderTab(tab, activeIdx) )}
				<Animated.View style={[tabStyle.tabUnderline, {
					width: this._itemWidth,
					// left: left
					left: this.state.aniLeft
				}, activeUnderlineStyle]} />
			</View>
		)

		return width > viewWidth ? (
			<View style={[{width: viewWidth}]} >
				<ScrollView
					horizontal
					/**
					 * 当值为true时，滚动条会停在滚动视图的尺寸的整数倍位置，可以用在水平分页上
					 * 默认值为false
					 * 此处为防止以后忘记以及加以提醒，还是显示的设置为 false
					 */
					pagingEnabled={false}
					automaticallyAdjustContentInsets={false}
					ref={s => {this._scrollView = s}}
					// onScrollBeginDrag={this._onScrollBeginDrag}
					onScrollEndDrag={this._onScrollEndDrag}
					// onMomentumScrollBegin={this._onMomentumScrollBegin}
					// onMomentumScrollEnd={this._onMomentumScrollEnd}
					showsHorizontalScrollIndicator={false}
					showsVerticalScrollIndicator={false}
					bounces={false}
					alwaysBounceVertical={false}
					alwaysBounceHorizontal={false}
				>
					{cnt}
				</ScrollView>
			</View>
		) : cnt
	}
}

const tabStyle = StyleSheet.create({
	tabUnderline: {
		position: 'absolute',
		height: 4,
		bottom: 0,
	}
})