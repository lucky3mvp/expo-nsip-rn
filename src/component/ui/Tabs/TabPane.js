import React, { Component } from 'react'
import { View, Text } from 'react-native'
export default class TabPane extends Component {
	render() {
		const { children, style={ flex: 1 },  ...rets } = this.props;
		return (
			<View style={style} {...rets}>
				{children}
			</View>
		);
	}
}