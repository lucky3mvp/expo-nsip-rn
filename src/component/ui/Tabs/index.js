/**
 * 标签页组件
 * @author esky 翟宇佳
 */
import React from 'react'
import PropTypes from 'prop-types'
import {
	Dimensions,
	View,
	Animated,
	ScrollView,
	StyleSheet,
	Platform,
	InteractionManager
} from 'react-native';
import TabBar from './TabBar';
import TabPane from './TabPane';
import UIComponent from '../../common/UIComponent';
import SceneComponent from './SceneComponent';

class Tabs extends UIComponent {
	static defaultProps = {
		renderTabBar: true,
		tabBarPosition: 'top',
		onChange: () => {},
		onTabClick: () => {},
		onAnimationEnd: () => {},
		animated: true,
		swipeable: true,
		compose: true,
		preRenderNum: 0,
	}

	static propTypes = {
		// 受控属性，当前激活 tab 面板的 key，需配合 onChange 维护 activeKey 值
		activeKey: PropTypes.oneOfType([
			PropTypes.string,
			PropTypes.number
		]),
		// 非受控属性，默认激活的 tab 面板的 key
		defaultActiveKey: PropTypes.oneOfType([
			PropTypes.string,
			PropTypes.number
		]),
		// 是否渲染 tabBar
		renderTabBar: PropTypes.bool,
		// tab 位置 top/bottom
		tabBarPosition: PropTypes.oneOf(['top', 'bottom']),
		// 切换面板（手动点击标签或是滑动 tab 内容切换）的回调
		onChange: PropTypes.func,
		// 点击 tab 标签的回调
		onTabClick: PropTypes.func,
		// tab 切换动画结束后执行的回调
		onAnimationEnd: PropTypes.func,
		// 是否显示切换动画
		animated: PropTypes.bool,
		// 是否可以滑动 tab 内容进行切换
		swipeable: PropTypes.bool,
		// 是否对内容进行压缩
		compose: PropTypes.bool,
		// 预加载的数量
		preRenderNum: PropTypes.number,
	}

	// 简单样式
	static simpleStyleProps = {
		textColor: { name: 'text', attr: 'color' },
		activeTextColor: { name: 'activeText', attr: 'color' },
		underlineColor: { name: 'bar', attr: 'borderColor' },
		activeUnderlineColor: { name: 'activeUnderline', attr: 'backgroundColor' },
	}

	static autoStyleSheet = false

	constructor(props) {
		super(props)

		let // 宽
			width = props.containerWidth || Dimensions.get("window").width,
			// 默认激活的 tab
			activeKey = props.activeKey === undefined ? props.defaultActiveKey : props.activeKey,
			// 激活 tab 的 index
			activeIdx = 0,
			// children
			{ children } = props,
			// 存放处理后的 tabs, 用于生成 tabBar
			tabs = [];

		// 预处理
		React.Children.forEach(children, (child, idx) => {
			// 找到初始显示的 tab
			if (activeKey === child.key) {
				activeIdx = idx;
			}
			
			tabs.push({
				_key: child.key, 
				index: idx,
				label: child.props.tab
			});
		});

		this.state = {
			containerWidth: width,
			curPageIdx: activeIdx,
			// _scrollX: new Animated.Value(activeIdx * width),
			// _aniScrollIdx: new Animated.Value(activeIdx)
		}

		this._showedSceneKeys = [activeIdx]
		this._tabs = tabs
		this._tabBarOffsetX = 0
	}

	componentDidMount() {
		this.mountGotoPageTimer = setTimeout( () => {
			this._scrollContent(this.state.curPageIdx, false);
			if(this.mountGotoPageTimer) clearTimeout(this.mountGotoPageTimer);
		}, 0)

		// InteractionManager.runAfterInteractions(() => {
		// 	this._scrollContent(this.state.curPageIdx, false);
		// })
	}

	componentWillReceiveProps(nextProps) {
		super.componentWillReceiveProps(nextProps)

		let activeKey = nextProps.activeKey;
		// 受控属性，由父元素告知 activeKey 来进行更新
		if (activeKey !== undefined && activeKey !== this.props.activeKey) {
			let idx = this._getIndex(activeKey);
			this._doAnimated(idx)
		}
	}

	componentWillUnmount() {
		if (this.handleLayoutTimer) clearTimeout(this.handleLayoutTimer)
		if (this.mountGotoPageTimer) clearTimeout(this.mountGotoPageTimer)
		// if(this.listener) this.state._scrollX.removeListener(this.listener)
	}

	_getTab(index) {
		let key = '', label = '';
		// 找到当前 index 对应的 tab
		this._tabs.forEach( (t, idx) => {
			if (t.index === index) {
				key = t._key,
				label = t.label
			}
		})

		return {
			key,
			label
		}
	}

	_getIndex(activeKey) {
		let index;
		// 找到 activeKey 对应的 tab index
		this._tabs.forEach( (t, idx) => {
			if (t._key === activeKey) {
				index = idx;
			}
		})

		return index;
	}

	_doAnimated(idx) {
		this.setState({
			curPageIdx: idx,
		}, () => {
			this._scrollContent(idx);
		});
	}

	_onTabBarScrollEnd = (offsetX) => {
		this._tabBarOffsetX = offsetX
	}

	_onTabClick = (key, label, index) => {
		let { activeKey, onChange, onTabClick } = this.props,
			{ curPageIdx } = this.state;

		// onTabClick 回调
		onTabClick(key, label);

		if (curPageIdx !== index) {
			// 使用的是非受控属性 defaultActiveKey，更新组件维护的值
			if (activeKey === undefined){
				this._doAnimated(index)
			} 
			// onChange 回调
			onChange(key, label);
		}
	}

	_scrollTabBar(nextPageIdx, animated) {
		let { containerWidth } = this.state,
			tbw = this.style.bar.width, // 若是滚动的 tab bar 则需更新 tab bar 位置
			tabBarWidth = tbw ? tbw : containerWidth,
			tabWidth = this.style.tab.width;

		if (tabBarWidth > containerWidth) {
			let start = tabWidth * nextPageIdx,
				end = tabWidth * (nextPageIdx + 1)

			if (start - this._tabBarOffsetX < 0) {
				this._tabBar._scrollView.scrollTo({x: start, y: 0, animated})
				this._tabBarOffsetX = start
			}
			if (end - this._tabBarOffsetX > containerWidth) {
				let offsetx = end - containerWidth
				this._tabBar._scrollView.scrollTo({x: offsetx, y: 0, animated})
				this._tabBarOffsetX = offsetx
			}
		}
	}

	_scrollContent = (nextPageIdx, animated = this.props.animated) => {
		let { containerWidth } = this.state,
			offset = nextPageIdx * containerWidth;

		this._updateSceneKeys(nextPageIdx);
		
		// 使用的是 Animated.ScrollView, 所以多了一层 _component
		// if (this._scrollView && this._scrollView._component && this._scrollView._component.scrollTo) {
		// 	this._scrollView._component.scrollTo({ x: offset, y: 0, animated });
		// 	this._scrollTabBar(nextPageIdx, animated);
		// }

		if (this._scrollView && this._scrollView.scrollTo) {
			this._scrollView.scrollTo({ x: offset, y: 0, animated });
			this._scrollTabBar(nextPageIdx, animated);
		}
	}

	_updateSceneKeys(nextPageIdx) {
		let newKeys = [], showedSceneKeys = this._showedSceneKeys;

		this._tabs.forEach( t => {
			if (showedSceneKeys.indexOf(t.index) >= 0 || this._shouldRenderSceneKey(t.index, nextPageIdx)) {
				newKeys.push(t.index);
			}
		})

		this._showedSceneKeys = newKeys
	}

	_shouldRenderSceneKey(idx, curPageIdx) {
		let numOfSibling = this.props.preRenderNum;

		return (idx < (curPageIdx + numOfSibling + 1) && idx > (curPageIdx - numOfSibling - 1));
	}

	_onMomentumScrollBeginAndEnd = (e) => {
		let offsetX = e.nativeEvent.contentOffset.x,
			{ containerWidth, curPageIdx } = this.state,
			{ activeKey, onChange } = this.props,
			index = Math.round(offsetX / containerWidth);

		if (curPageIdx !== index) {
			// 使用的是非受控属性 defaultActiveKey，更新组件维护的值
			if (activeKey === undefined){
				this._doAnimated(index)
			} 
			
			let { key, label } = this._getTab(index);
			onChange(key, label);
		}
	}

	_onAnimationEnd = (activeIdx) => {
		let { key, label } = this._getTab(activeIdx);
		this.props.onAnimationEnd(key, label);
	}

	_renderTabBar = () => {
		if (this.props.renderTabBar === false) return null;

		let style = this.style,
			{ containerWidth, curPageIdx } = this.state,
			{ tabBarPosition } = this.props,
			tabBarTop = tabBarPosition === 'top',
			barBaseStyle = tabBarTop ? style.barTop : style.barBottom,
			linePosition = tabBarTop ? {} : { top: -1 },
			activeUnderlineStyle = [style.activeUnderline, {
				bottom: tabBarTop ? -1 : null,
			}, linePosition];

		let tabBarProps = {
			// aniScrollIdx: this.state._aniScrollIdx,
			tabs: this._tabs,
			activeIdx: curPageIdx,
			tabWidth: style.tab.width, // 单个 tab 宽
			containerWidth: style.bar.width, // tabBar 容器宽
			viewWidth: containerWidth, // 内容区宽
			onTabClick: this._onTabClick,
			onScrollEnd: this._onTabBarScrollEnd,
			onAnimationEnd: this._onAnimationEnd
		};

		return (
			<TabBar
				ref={t => this._tabBar = t}
				style={[style.bar, barBaseStyle]}
				textStyle={style.text}
				activeTextStyle={style.activeText}
				tabStyle={style.tab}
				activeUnderlineStyle={activeUnderlineStyle}
				{...tabBarProps}
			/>
		);
	}

	_composeScenes() {
		let { containerWidth, curPageIdx } = this.state, 
			{ compose } = this.props,
			showedSceneKeys = this._showedSceneKeys;

		return this.props.children.map( (child, idx) => {
			return (
				<SceneComponent
					key={idx}
					shouldUpdated={this._shouldRenderSceneKey(idx, curPageIdx)}
					style={{width: containerWidth}}
				>
				{
					compose ? (
						(showedSceneKeys.indexOf(idx) >= 0) || (idx === curPageIdx) ? child : <View tab={child.props.tab}/>
					) : (
						child
					)
				}
				</SceneComponent>
			)
		})
	}

	_onScroll = (e) => {
		// this.state._scrollX.setValue(e.nativeEvent.contentOffset.x)
	}

	_renderContent() {
		let { curPageIdx, containerWidth } = this.state,
			{ swipeable } = this.props;

		return (
			<ScrollView
				keyboardShouldPersistTaps='handled'
				keyboardDismissMode="on-drag"
				horizontal
				pagingEnabled
				automaticallyAdjustContentInsets={false}
				ref={s => {this._scrollView = s}}
				// onScroll={Animated.event(
				// 	[{
				// 		nativeEvent: { contentOffset: { x: this.state._scrollX } }
				// 	}], {
				// 		useNativeDriver: true,
				// 	})}
				onScroll={this._onScroll}
				onMomentumScrollBegin={this._onMomentumScrollBeginAndEnd}
				onMomentumScrollEnd={this._onMomentumScrollBeginAndEnd}
				scrollEventThrottle={16}
				scrollsToTop={false}
				showsHorizontalScrollIndicator={false}
				scrollEnabled={swipeable}
				directionalLockEnabled
				alwaysBounceVertical={false}
			>
				{this._composeScenes()}
			</ScrollView>
		);
	}

	_scrollXListener = (containerWidth, {value}) => {
		// this.state._aniScrollIdx.setValue((value / containerWidth));

		// /**
		//  * ios 里会出现触发两次 value 一样的回调，所以加以区别 
		//  */
		// if ((this._prevValue !== value) && (value === containerWidth * this.state.curPageIdx)) {
		// 	this._prevValue = value;
		// 	this.props.onAnimationEnd();
		// }
	}

	_handleLayout = (e) => {
		const { width } = e.nativeEvent.layout,
			{ curPageIdx, containerWidth } = this.state;
		let finalWidth = containerWidth;

		if (Math.round(width) !== Math.round(containerWidth)) {
			finalWidth = width;

			// this.state._scrollX.setValue(curPageIdx * width);
			// this.state._aniScrollIdx.setValue(curPageIdx);

			this.setState({containerWidth: width});
			
			if (this.handleLayoutTimer) clearTimeout(this.handleLayoutTimer);
			this.handleLayoutTimer = setTimeout(() => {
				this._scrollContent(curPageIdx, false);
				if (this.handleLayoutTimer) clearTimeout(this.handleLayoutTimer);
			}, 0);


			// InteractionManager.runAfterInteractions(() => {
			// 	this._scrollContent(curPageIdx, false);
			// })
		}

		// if(this.listener) this.state._scrollX.removeListener(this.listener);
		// this.listener = this.state._scrollX.addListener(this._scrollXListener.bind(this, finalWidth));
	}

	render() {
		let { tabBarPosition } = this.props;
		return (
			<View onLayout={this._handleLayout} style={[{flex: 1}, this.style.container]}>
				{tabBarPosition === 'top' && this._renderTabBar()}
				{this._renderContent()}
				{tabBarPosition === 'bottom' && this._renderTabBar()}
			</View>
		);
	}
}

Tabs.baseStyle = {
	// 单个 tab 标签
	tab: {
		flex: 1,
		paddingBottom: 0,
		alignItems: 'center',
		justifyContent: 'center',
	},
	// tab 标签内文字
	text: {
		fontSize: 15,
		color: '#333'
	},
	// 激活态 tab 标签内文字
	activeText: {
		color: '#108ee9'
	},
	// tab bar
	bar: {
		height: 42,
		flexDirection: 'row',
		borderColor: '#ddd'
	},
	barTop: {
		borderTopWidth: 0,
		borderBottomWidth: StyleSheet.hairlineWidth,
	},
	barBottom: {
		borderTopWidth: StyleSheet.hairlineWidth,
		borderBottomWidth: 0,
	},
	// 激活态线条
	activeUnderline: {
		height: 2,
		backgroundColor: '#108ee9'
	},
}

Tabs.TabPane = TabPane
export default Tabs