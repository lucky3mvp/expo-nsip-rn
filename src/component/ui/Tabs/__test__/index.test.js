import React from 'react'
import Renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import Tabs from '../index'
import TabBar from '../../Tabs/TabBar.js'

describe('Tabs Tests', () => {
	let handler;
	let content;

	beforeEach(() => {
		handler = jest.fn();
		content = [];
		let tabs = ['tab1', 'tab2', 'tab3'];
		for (let key in tabs) {
			let tab = tabs[key];
			content.push(
				<Tabs.TabPane tab={tab} key={tab} />
			)
		}
	});
	
	it('renders correctly', () => {
		const tree = Renderer.create((
			<Tabs>
				{content}
			</Tabs>
		)).toJSON();
		
		expect(tree).toMatchSnapshot();
	});

	it('shallow function', () => {
		const wrapper = shallow(
			<Tabs defaultActiveKey='0'>
				{content}
			</Tabs>
		);
		
		expect(wrapper.find(TabBar)).toHaveLength(1);
	});
})
