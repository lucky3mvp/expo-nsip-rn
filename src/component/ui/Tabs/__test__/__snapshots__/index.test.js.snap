// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs Tests renders correctly 1`] = `
<View
  onLayout={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "borderColor": "#ddd",
          "flexDirection": "row",
          "height": 42,
        },
        Object {
          "borderBottomWidth": 0.5,
          "borderTopWidth": 0,
        },
      ]
    }
  >
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel="tab1"
      accessibilityTraits="button"
      accessible={true}
      hitSlop={undefined}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
          "paddingBottom": 0,
        }
      }
      testID={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#333",
              "fontSize": 15,
            },
            Object {
              "color": "#108ee9",
            },
            Object {
              "fontWeight": "bold",
            },
          ]
        }
      >
        tab1
      </Text>
    </View>
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel="tab2"
      accessibilityTraits="button"
      accessible={true}
      hitSlop={undefined}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
          "paddingBottom": 0,
        }
      }
      testID={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#333",
              "fontSize": 15,
            },
            Object {},
            Object {
              "fontWeight": "normal",
            },
          ]
        }
      >
        tab2
      </Text>
    </View>
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel="tab3"
      accessibilityTraits="button"
      accessible={true}
      hitSlop={undefined}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
          "paddingBottom": 0,
        }
      }
      testID={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Array [
            Object {
              "color": "#333",
              "fontSize": 15,
            },
            Object {},
            Object {
              "fontWeight": "normal",
            },
          ]
        }
      >
        tab3
      </Text>
    </View>
    <View
      collapsable={undefined}
      style={
        Object {
          "backgroundColor": "#108ee9",
          "bottom": -1,
          "height": 2,
          "left": 0,
          "position": "absolute",
          "width": 250,
        }
      }
    />
  </View>
  <RCTScrollView
    alwaysBounceVertical={false}
    automaticallyAdjustContentInsets={false}
    directionalLockEnabled={true}
    horizontal={true}
    keyboardDismissMode="on-drag"
    keyboardShouldPersistTaps="handled"
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    pagingEnabled={true}
    scrollEnabled={true}
    scrollEventThrottle={16}
    scrollsToTop={false}
    showsHorizontalScrollIndicator={false}
  >
    <View>
      <View
        style={
          Object {
            "width": 750,
          }
        }
      >
        <View
          style={
            Object {
              "flex": 1,
            }
          }
          tab="tab1"
        />
      </View>
      <View
        style={
          Object {
            "width": 750,
          }
        }
      >
        <View
          tab="tab2"
        />
      </View>
      <View
        style={
          Object {
            "width": 750,
          }
        }
      >
        <View
          tab="tab3"
        />
      </View>
    </View>
  </RCTScrollView>
</View>
`;
